### ISSUE TITLE
Idea: Erweiterte Spaced Repetition Algorithmen (SM-17/18, FSRS, ML-basiert)

### ISSUE THREAD

**Issue by trsdn** [trsdn/learning-platform#19]
## Feature Idea: Erweiterte Spaced Repetition Algorithmen

Upgrade from SM-2 to modern spaced repetition algorithms for more accurate review scheduling and better learning outcomes.

### Current State: SM-2 (1987)

**Limitations**:
- Fixed difficulty scaling
- No personalization beyond EF (easiness factor)
- Doesn't account for individual learning patterns
- Simple exponential intervals
- No prediction of optimal review time

#### 1. SM-17/18 (SuperMemo 17/18)

**Improvements over SM-2**:
- Two-component model of memory
- Retrievability and stability tracking
- More granular difficulty assessment
- Better long-term retention prediction

**Features**:
- Adaptive difficulty based on retrievability
- Optimal interval calculation using forgetting curves
- Considers item difficulty AND user ability

**Implementation Complexity**: Medium

#### 2. FSRS (Free Spaced Repetition Scheduler) - 2024

**What is FSRS**:
- Open-source alternative to SM-17/18
- Based on research from 2024
- Free to use (no licensing)
- Proven better retention than SM-2

**Key Features**:
- Four parameters per card
- Memory stability tracking
- Difficulty estimation
- Optimal retention rate balancing

**Algorithm**:
```typescript
interface FSRSCard {
  due: Date;
  stability: number;      // Memory stability
  difficulty: number;     // Item difficulty (0-10)
  elapsedDays: number;
  scheduledDays: number;
  reps: number;
  lapses: number;
}

class FSRS {
  // Parameters (can be optimized per user)
  private w = [0.4, 0.6, 2.4, 5.8, 4.93, 0.94, 0.86, 0.01, 1.49, 0.14, 0.94, 2.18, 0.05, 0.34, 1.26, 0.29, 2.61];
  
  schedule(card: FSRSCard, rating: 1 | 2 | 3 | 4): FSRSCard {
    const interval = this.calculateInterval(card, rating);
    return {
      ...card,
      due: addDays(new Date(), interval),
      stability: this.nextStability(card, rating),
      difficulty: this.nextDifficulty(card, rating),
      reps: card.reps + 1
    };
  }
  
  private calculateInterval(card: FSRSCard, rating: number): number {
    // Complex calculation based on stability and difficulty
    // See: https://github.com/open-spaced-repetition/fsrs4anki
  }
}
```

**Benefits**:
- 10-20% better retention than SM-2
- Fewer reviews for same retention
- Adapts to user's learning ability
- Free and open-source

**Implementation Complexity**: Medium-High

#### 3. Neural Network Based Scheduling

**ML Approach**:
- Train neural network on user's review history
- Predict optimal review time per user
- Personalized to individual learning patterns
- Continuous learning and improvement

**Features**:
- User-specific models
- Considers context (time of day, device, topic)
- Predicts forgetting curve per card
- Adaptive to changing learning ability

**Architecture**:
```typescript
interface NeuralScheduler {
  // Input features
  features: {
    cardHistory: ReviewHistory[];
    userPatterns: LearningPattern;
    currentPerformance: number;
    timeContext: TimeContext;
  };
  
  // Model
  model: TensorFlowModel;
  
  // Prediction
  predictNextReview(card: Card): Date;
  predictDifficulty(card: Card): number;
  updateModel(actualResult: ReviewResult): void;
}
```

**Training Data**:
- User's review history (>100 reviews)
- Performance metrics
- Time patterns
- Forgetting curves

**Benefits**:
- Highest potential accuracy
- Truly personalized
- Learns from user's specific patterns
- Can optimize for time efficiency or retention

**Implementation Complexity**: High

### Comparison Table

| Algorithm | Retention | Reviews Needed | Personalization | Complexity | Open Source |
|-----------|-----------|----------------|-----------------|------------|-------------|
| SM-2      | Baseline  | Baseline       | Low             | Low        | Yes         |
| SM-17/18  | +15%      | -10%           | Medium          | Medium     | No (paid)   |
| FSRS      | +20%      | -15%           | High            | Medium     | Yes         |
| Neural    | +25%      | -20%           | Highest         | High       | Depends     |

### Suggested Implementation Strategy

**Phase 1: FSRS Integration**
- Implement FSRS algorithm
- Migrate existing SM-2 data
- A/B test against SM-2
- Measure retention improvement

**Phase 2: User Choice**
- Allow users to select algorithm
- Provide algorithm comparison stats
- Settings: SM-2, FSRS, Auto (best for user)

**Phase 3: ML Enhancement (Future)**
- Collect anonymized review data
- Train neural models
- Offer ML-based scheduling (beta)
- Continuous model improvement

### Migration Strategy

**Data Migration**:
```typescript
// Migrate SM-2 to FSRS
function migrateToFSRS(sm2Card: SM2Card): FSRSCard {
  return {
    due: sm2Card.nextReview,
    stability: estimateStability(sm2Card),
    difficulty: mapEFtoDifficulty(sm2Card.easinessFactor),
    elapsedDays: 0,
    scheduledDays: sm2Card.interval,
    reps: sm2Card.repetitions,
    lapses: 0
  };
}

function estimateStability(sm2Card: SM2Card): number {
  // Estimate from interval and EF
  return sm2Card.interval * sm2Card.easinessFactor;
}
```

**Backward Compatibility**:
- Keep SM-2 as fallback
- Allow algorithm switching
- Preserve all historical data

### User Settings

**Algorithm Selection**:
```
⚙️ Lernalgorithmus

Wähle deinen Wiederholungsalgorithmus:

( ) SM-2 (Standard)
    Bewährt seit 1987, einfach und zuverlässig

(•) FSRS (Empfohlen)
    Modern, 20% bessere Retention, weniger Wiederholungen

( ) Neural (Beta)
    KI-basiert, lernt von deinem Verhalten
    Benötigt >100 Wiederholungen für Aktivierung

📊 Statistiken: 
- Deine Retention: 87%
- Vorhergesagte Verbesserung mit FSRS: +12%
```

### Benefits

**For Students**:
- Better retention with fewer reviews
- Optimized review scheduling
- Personalized to learning style
- More efficient learning

**For Platform**:
- Competitive with Anki, Quizlet
- Modern algorithm offering
- Better user outcomes
- Differentiation feature

### Technical Considerations

**FSRS Implementation**:
- TypeScript port of FSRS
- Unit tests against reference implementation
- Performance optimization
- IndexedDB schema update

**ML Implementation** (Future):
- TensorFlow.js for client-side inference
- Privacy-preserving training (federated learning)
- Model versioning and updates
- Fallback to FSRS if model unavailable

**Testing**:
- Simulate 1000+ review sessions
- Compare retention rates
- Measure review frequency
- A/B test with real users

### Research References

- **SM-2**: Wozniak, P. (1990). "Algorithm SM-2"
- **SM-17**: SuperMemo Algorithm Wiki
- **FSRS**: Jarrett Ye (2024). "Free Spaced Repetition Scheduler"
- **Neural**: Settles & Meeder (2016). "ML for Spaced Repetition"

### Related to

- Current SM-2 implementation
- Progress analytics dashboard
- #18 (AI Integration - can suggest optimal algorithm)
- Performance optimization

### Implementation Checklist

**FSRS**:
- [ ] Port FSRS algorithm to TypeScript
- [ ] Create FSRS card interface
- [ ] Implement scheduling logic
- [ ] Migration from SM-2
- [ ] Unit tests (100+ test cases)
- [ ] A/B testing framework
- [ ] User settings UI
- [ ] Documentation

**ML** (Future):
- [ ] Design neural architecture
- [ ] Implement TensorFlow.js integration
- [ ] Training pipeline
- [ ] Privacy-preserving learning
- [ ] Model evaluation metrics
- [ ] Fallback mechanisms

### External Resources

- FSRS Implementation: https://github.com/open-spaced-repetition/fsrs4anki
- Research Paper: https://arxiv.org/abs/2401.14605
- Benchmark Results: https://github.com/open-spaced-repetition/srs-benchmark
